File: 383-ransom-note/383-ransom-note.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
  const N = magazine.length,
        M = ransomNote.length,
        dict = new Array<number>(26).fill(0);
  
  for(let i = 0; i < N; i++) {
    let idx = magazine.charCodeAt(i) - 97;
    dict[idx] += 1;
  }

  for(let i = 0; i < M; i++) {
    let idx = ransomNote.charCodeAt(i) - 97;
    dict[idx] -= 1;
    
    if(dict[idx] < 0)
      return false;
  }
  
  return true;
};

File: 383-ransom-note/383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] dict = new int[26];
        
        for(char c: magazine.toCharArray())
            dict[c-'a'] += 1;
        
        for(char rn: ransomNote.toCharArray()) 
            if(dict[rn-'a'] == 0)
                return false;
            else
                dict[rn-'a'] -= 1;
        
        return true;
    }
}